From f343e5d26631888170d914b444558cec8139b294 Mon Sep 17 00:00:00 2001
From: Patrick Williams <patrick@stwcx.xyz>
Date: Thu, 27 Feb 2025 18:39:58 -0500
Subject: [PATCH 200/205] registries: generate registries_selector.hpp

Enhance parse_registries.py to generate the registries_selector.hpp
so that it does not need to be generated by hand.

Note that as part of this, additional registries are now emitted for
Redfish.  These registries were already being generated but they were
missing the header hook that allowed them to be exposed.

Tested:
```
$ curl -s -k https://localhost:18080/redfish/v1/Registries/ | jq '.Members | map(."@odata.id")'
[
  "/redfish/v1/Registries/Base",
  "/redfish/v1/Registries/Composition",
  "/redfish/v1/Registries/Environmental",
  "/redfish/v1/Registries/EthernetFabric",
  "/redfish/v1/Registries/Fabric",
  "/redfish/v1/Registries/HeartbeatEvent",
  "/redfish/v1/Registries/JobEvent",
  "/redfish/v1/Registries/License",
  "/redfish/v1/Registries/LogService",
  "/redfish/v1/Registries/NetworkDevice",
  "/redfish/v1/Registries/OpenBMC",
  "/redfish/v1/Registries/Platform",
  "/redfish/v1/Registries/Power",
  "/redfish/v1/Registries/ResourceEvent",
  "/redfish/v1/Registries/SensorEvent",
  "/redfish/v1/Registries/StorageDevice",
  "/redfish/v1/Registries/TaskEvent",
  "/redfish/v1/Registries/Telemetry",
  "/redfish/v1/Registries/Update"
]
```

Signed-off-by: Patrick Williams <patrick@stwcx.xyz>
Change-Id: Ia243a7bfbb8873384b6679b462914ba8380bf5b3
---
 redfish-core/include/registries_selector.hpp | 23 +++++++++++++-------
 scripts/parse_registries.py                  | 21 ++++++++++++++++++
 2 files changed, 36 insertions(+), 8 deletions(-)

diff --git a/redfish-core/include/registries_selector.hpp b/redfish-core/include/registries_selector.hpp
index 6440ca59..71a628db 100644
--- a/redfish-core/include/registries_selector.hpp
+++ b/redfish-core/include/registries_selector.hpp
@@ -1,17 +1,24 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-FileCopyrightText: Copyright OpenBMC Authors
 #pragma once
-// NOLINTNEXTLINE(misc-include-cleaner)
-#include "registries.hpp"
-// NOLINTNEXTLINE(misc-include-cleaner)
+// NOLINTBEGIN(misc-include-cleaner)
 #include "registries/base_message_registry.hpp"
-// NOLINTNEXTLINE(misc-include-cleaner)
+#include "registries/composition_message_registry.hpp"
+#include "registries/environmental_message_registry.hpp"
+#include "registries/ethernet_fabric_message_registry.hpp"
+#include "registries/fabric_message_registry.hpp"
 #include "registries/heartbeat_event_message_registry.hpp"
-// NOLINTNEXTLINE(misc-include-cleaner)
+#include "registries/job_event_message_registry.hpp"
+#include "registries/license_message_registry.hpp"
+#include "registries/log_service_message_registry.hpp"
+#include "registries/network_device_message_registry.hpp"
 #include "registries/openbmc_message_registry.hpp"
-// NOLINTNEXTLINE(misc-include-cleaner)
+#include "registries/platform_message_registry.hpp"
+#include "registries/power_message_registry.hpp"
 #include "registries/resource_event_message_registry.hpp"
-// NOLINTNEXTLINE(misc-include-cleaner)
+#include "registries/sensor_event_message_registry.hpp"
+#include "registries/storage_device_message_registry.hpp"
 #include "registries/task_event_message_registry.hpp"
-// NOLINTNEXTLINE(misc-include-cleaner)
 #include "registries/telemetry_message_registry.hpp"
+#include "registries/update_message_registry.hpp"
+// NOLINTEND(misc-include-cleaner)
diff --git a/scripts/parse_registries.py b/scripts/parse_registries.py
index 03e44cb3..595d9a83 100755
--- a/scripts/parse_registries.py
+++ b/scripts/parse_registries.py
@@ -87,7 +87,28 @@ def openbmc_local_getter() -> RegistryInfo:
     return (path, json_file, "openbmc", url)
 
 
+def generate_selector(files: t.List[RegistryInfo]) -> None:
+    # Create the top-level header.
+    with open(
+        os.path.join(
+            SCRIPT_DIR,
+            "..",
+            "redfish-core",
+            "include",
+            "registries_selector.hpp",
+        ),
+        "w",
+    ) as selector:
+        selector.write(f"{COPYRIGHT}{PRAGMA_ONCE}")
+        selector.write("// NOLINTBEGIN(misc-include-cleaner)\n")
+        for file in sorted([os.path.basename(x[0]) for x in files]):
+            selector.write(f'#include "registries/{file}"\n')
+        selector.write("// NOLINTEND(misc-include-cleaner)\n")
+
+
 def update_registries(files: t.List[RegistryInfo]) -> None:
+    generate_selector(files)
+
     # Remove the old files
     for file, json_dict, namespace, url in files:
         try:
-- 
2.47.1

